### ISSUE TITLE
New Select Box not behaving

### ISSUE THREAD

**Issue by glideranderson** [revsystems/jQuery-SelectBox#5]
'http://glideint.com/clients/beertoyou/homeBrew?sort=pd.name&order=DESC'

There is a sort select field that I want to recreate via your plugin. However, the functionality doesn't stay the same. The select box has an onchange = "location = this.value". I do not know how to recreate this functionality with your js. With your plugin active it just refreshes the page without changing the sort order.

This is an awesome plugin though!

**Comment by MaxPower15** [comment]
Hi,

Looking at your select, I see you're putting complete URLs in the value field for options. I think jquery-sb is struggling because it stores the values in the markup, and some of them (?, =, etc.) are probably encoded by default. So when jquery-sb tries to match the underlying select to the replacement, it can't find a match.

My quick solution for you would be to encode the URLs that you put in the option values. When this is done, the value would look like something this:

http%3A%2F%2Fglideint.com%2Fclients%2Fbeertoyou%2FhomeBrew%3Fsort%3Dp.sort_order%26order%3DASC

 I'm not sure what language you're using, but it's urlencode() in php--I'm sure there's something similar in what you're using.

After they're encoded, I think the javascript will work properly. If it has issues with the encoded string, you can use window.location = decodeURI(this.value); in the javascript. That should fix it.

I will look into automatically encoding/decoding these values in a newer release.

Thanks for the report!

**Comment by glideranderson** [comment]
Thank you for your amazingly quick response! I encoded my urls with the urlencode() (using php) before being added to the select field. I also added the window.location with the decodeURI function to the select onchange event. Now the pages are not being found. I know I'm missing something. I love this plugin (so far my favorite) and it helps make my site great. Can I get so help making this work (it's in Opencart so I haven't messed with the core functionality much)?

http://glideint.com/clients/beertoyou/homeBrew

Thank you, thank you. Very much appreciate it.

**Comment by MaxPower15** [comment]
Okay, I just pushed a new version of jquery-sb that should fix your issue. Just drop the new js in, set your code back to the way it was, and I think it will work. Turned out to be a quick fix on my end.

BTW, a quick look at the HTML and I think your issue is that only the first option was being urlencoded. The others were still normal.

Glad you like jquery-sb. Hopefully this gets it working! ; )

**Comment by glideranderson** [comment]
Thank you! It works wonderful!!

**Comment by glideranderson** [comment]
I spoke too soon. Now when in IE 8 and IE 7, when I click on the select field I get redirected to my home page. The drop down never displays.

**Comment by MaxPower15** [comment]
Something's definitely wrong with those. Your problem is as described, but on my test page, IE is straight-up crashing. Bleh. I'll try to figure out what's wrong.

Thanks for keeping on it.

**Comment by MaxPower15** [comment]
This seems like a versioning issue of some kind. My tests with jquery versions > 1.4.1 work in all versions of IE. Below that, it crashes IE7 and IE8. If you're able to upgrade to that, feel free. I'll try to get some kind of fix down to 1.3.x out soon though.

**Comment by MaxPower15** [comment]
Okay, I fixed a pretty annoying bug in IE<9+jquery<1.4.2. I'm not sure if this is related to your issue, however.

For your issue, when I click the box, I'm routed somewhere else. When I mess with IE9's debugger (I'm testing IE7/8 using IE9 different browser modes) a bit to show the options, however, I can select the option and it goes to the correct location.

Since the display has href="#" it shouldn't never really get rid of all your URL parameters. At worst (if I wasn't preventing the default) it should append it to the end without refreshing. I also duplicated your select markup using jquery 1.3.2 on the demo page, but it appeared to be working correctly there.

In short: I think you have some other javascript in play that's redirecting you when you click the sb display. I don't know where it's coming from though.

Hope that helps a bit.

**Comment by MaxPower15** [comment]
This issue was bothering me so I spent some more time debugging in IE. The problem appears to be that an error is thrown when the selectbox is positioning itself. I think it is dying because the margin-left on the body element is set to "auto", which causes an exception somewhere down in jquery's internals, probably when it tries to convert it to an integer. But I was unable to replicate the error and it's hard to debug jquery when it's minified, so I could be a little off.

The code to position based on the body margin exists to fix a different bug (maybe jquery? maybe browser? i'm not sure)--basically, if you set a margin on the body, jquery's offset() function no longer works correctly and the dropdown would always appear shifted.

Looking at your markup and css, I see you're using margin:0 auto; on the body element. And sure enough, I just tried changing the css on body from margin:0px auto to margin:0 0, and jquery-sb started working. Your site didn't appear any different.

Hope that fixes it for you!

**Comment by revsystems** [comment]
I haven't heard anything on this in a while, so I'm going to close.

I introduced some version-specific code a while back to get the body margins without throwing an error. Hopefully that handles the problem in a wider sense.
